### PR TITLE
Implement struct terms and eliminations

### LINKED ISSUE #247: Struct terms and eliminations
For example, given the following `Point` struct:

```
struct Point : Type {
    x : Int,
    y : Int,
}
```

We'll want to be able to construct a term of that type, and project on its fields:

```
origin : Point = struct {
    x = 0,
    y = 0,
};
origin_x = origin.x;
origin_y = origin.y;
```

Add ability to construct terms of the type `struct` that we are able to introduce, as well as provide access to fields.

### PR DESCRIPTION
**Pull request by brendanzab**
This implements struct terms and eliminations. For example, we can now write:

```
struct Point : Type {
    x : Int,
    y : Int,
}

const origin : Point = struct {
    x = 0,
    y = 0,
};

const origin_x = origin.x;
const origin_y = origin.y;
```

Closes #247.

### CODE REVIEW

File: book/src/reference/structs.md
Comment (toothbrush) [inline book/src/reference/structs.md:75]: Re: not forgetting the context of `MyArray` which contains `len` and `data`, they need to remain correctly lexically scoped so that when you try to use `array.data` down below, you're able to correctly infer/check that it has type `Array len Point`.

Is it worth creating an issue about this so we don't forget?

Comment (brendanzab) [reply book/src/reference/structs.md:75]: Yeah, we could add it to #244 perhaps?

File: fathom/src/lang/core/typing.rs
Comment (toothbrush) [inline fathom/src/lang/core/typing.rs:176]: From discussion: is this `FIXME` still a thing?

File: fathom/src/pass/surface_to_core.rs
Comment (toothbrush) [inline fathom/src/pass/surface_to_core.rs:167]:
```
                         let item_data = match self.eval(&core_type).as_ref() { 
```

Let's not redo evaluation of the `core_type`.

File: tests/input/fail_duplicate_definitions.fathom
Comment (toothbrush) [inline tests/input/fail_duplicate_definitions.fathom:4]: Maybe it'd be helpful to have the user write `const Empty = 1 : Int;` when introducing this binding, otherwise it reads a bit like defining the type of `Empty` as `Format` and subsequently assigning a value to it.

Comment (brendanzab) [reply tests/input/fail_duplicate_definitions.fathom]: Made an issue for this here: #250

Comment (toothbrush) [reply tests/input/fail_duplicate_definitions.fathom]: Cool, thanks – i like being organised with cards we can work on :nerd_face:

**Review by toothbrush — APPROVED** [review]
Apologies for the delay, but i finally had a look through this.  From what i can see it looks very reasonable, thanks for cleaning up the error handling i didn't manage to do.

I don't feel like i can really speak to the specifics, but if you're waiting for my blessing, i'd say merge away! :+1: :shipit:

**Comment by toothbrush** [comment]
Merging after discussion!

**Pull request merged**
